### PR TITLE
Miscellaneous project system decoupling

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This is a series of random commits I've made as I've been doing the big refactoring for the free-threaded project system work. Commit-by-commit highly recommended as these are entirely unrelated changes. :-) Just making this pull request to merge these rather than having to rebase them all the time.

General philosophy here was code that that was consuming AbstractProject only to get something you could use more directly has been moved to the "more direct" option.

<details><summary>Ask Mode template</summary>

### Customer scenario

This is miscellaneous cleanup that's part of the larger work to move our language service initialization to a background thread.

### Bugs this fixes

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/501654

### Workarounds, if any

Nope.

### Risk

Very low.

### Performance impact

None.

</details>

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 1]
```diff
@@ -493,6 +493,22 @@ protected internal void OnOutputFilePathChanged(ProjectId projectId, string outp
             }
         }
 
+        /// <summary>
+        /// Call this method when a project's output ref file path is changed in the host environment.
+        /// </summary>
+        protected internal void OnOutputRefFilePathChanged(ProjectId projectId, string outputFilePath)
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs]: Is this added speculatively, or should there be a caller already?  This change seems a bit different from the rest of the PR.

File: src/VisualStudio/Core/Def/Implementation/EditAndContinue/VsENCRebuildableProjectImpl.cs [hunk 2]
```diff
@@ -169,17 +169,7 @@ internal bool OnEdit(DocumentId documentId)
...
-                        if (deferredLoad)
-                        {
-                            message = ServicesVSResources.ChangesNotAllowedIfProjectWasntLoadedWhileDebugging;
-                            s_encDebuggingSessionInfo?.LogReadOnlyEditAttemptedProjectNotBuiltOrLoaded();
-                        }
-                        else
-                        {
-                            message = ServicesVSResources.ChangesNotAllowedIfProjectWasntBuildWhenDebuggingStarted;
-                        }
+                        message = ServicesVSResources.ChangesNotAllowedIfProjectWasntBuildWhenDebuggingStarted;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/EditAndContinue/VsENCRebuildableProjectImpl.cs]: nit.  the string is "Built" but the identifier says "Build". (not actually related to your change).

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs]: I'm using it in my uber-commit with the big refactoring.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please.

**Comment by jasonmalinowski** [comment]
Tagging @jinujoseph and @Pilchie for ask mode approval.

**Comment by Pilchie** [comment]
Approved for 15.8 Preview 2.

**Pull request merged**